### PR TITLE
Pseudo Labeling Distillation

### PR DESCRIPTION
**Pull request by shira-g**

### CODE REVIEW

File: tests/test_pseudo_label_distillation.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+from nlp_architect.data.sequential_tagging import TokenClsProcessor
+
+
+def test_distillation():
+    current_dir = os.path.dirname(os.path.realpath(__file__))
+    data_dir = os.path.join(current_dir, 'fixtures/data/distillation')
+    num_of_examples = count_examples(data_dir + os.sep + 'train.txt')
+    labeled_precentage = 0.4
+    unlabeled_precentage = 0.5
+    bert_output_path = data_dir + os.sep + 'bert_out'
```
Comment (peteriz) [inline tests/test_pseudo_label_distillation.py]: switch to temp directory (python can provide that)

File: tests/test_pseudo_label_distillation.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    kd_output_path = data_dir + os.sep + 'kd_out'
```
Comment (peteriz) [inline tests/test_pseudo_label_distillation.py]: same

File: tests/test_pseudo_label_distillation.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    assert(labeled_precentage <= 1 and unlabeled_precentage <= 1)
+
+    # test split dataset
```
Comment (peteriz) [inline tests/test_pseudo_label_distillation.py]: please write a separate test for each functionality that you're testing

Comment (peteriz) [reply tests/test_pseudo_label_distillation.py]: the split function should be in a separate test file (its a data method)

File: tests/test_pseudo_label_distillation.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        processor.split_dataset(
+            dataset='train.txt', labeled=math.ceil(num_of_examples * labeled_precentage),
+            unlabeled=math.ceil(num_of_examples * unlabeled_precentage), out_folder=data_dir)
+        check_labeled_count = count_examples(data_dir + os.sep + 'labeled.txt')
+        assert(check_labeled_count == math.ceil(num_of_examples * labeled_precentage))
+        check_unlabeled_count = count_examples(data_dir + os.sep + 'unlabeled.txt')
+        assert(check_unlabeled_count == math.ceil(num_of_examples * unlabeled_precentage))
+
+    # test bert ner
+    parser = argparse.ArgumentParser()
```
Comment (peteriz) [inline tests/test_pseudo_label_distillation.py]: you should test the model instead of the procedure

File: tests/test_pseudo_label_distillation.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        ['--data_dir', data_dir, '--output_dir',
+            bert_output_path, '--model_name_or_path',
+            'bert-base-cased', '--model_type', 'bert',
+            '--train_file_name', 'labeled.txt', '--num_train_epoch',
+            '1', '--per_gpu_train_batch_size', '50', '--overwrite_output_dir'])
+    train_proc.run_procedure(args)
+    assert(os.path.exists(bert_output_path + os.sep + 'pytorch_model.bin'))
+
+    # test distillation pseudo
+    parser = argparse.ArgumentParser()
```
Comment (peteriz) [inline tests/test_pseudo_label_distillation.py]: same as above

File: tests/test_pseudo_label_distillation.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        ['--data_dir', data_dir, '--output_dir', out_dir,
+            '--teacher_model_path', bert_output_path,
+            '--teacher_model_type', 'bert', '--labeled_train_file',
+            'labeled.txt', '--unlabeled_train_file', 'unlabeled.txt',
+            '-e', '10', '--overwrite_output_dir'])
+    train_proc.run_procedure(args)
+    assert(os.path.exists(out_dir + os.sep + 'model.bin'))
+
+    # delete test files
+    if os.path.exists(bert_output_path):
```
Comment (peteriz) [inline tests/test_pseudo_label_distillation.py]: this should be removed once you move to temp directories

File: nlp_architect/data/sequential_tagging.py [hunk 7]
```diff
@@ -355,6 +356,25 @@ def get_labels(self):
     def get_labels_filename():
         return "labels.txt"
 
+    def split_dataset(self, labeled: int, unlabeled: int, out_folder, dataset: str = 'train.txt'):
```
[the remark this answers is not in the document]
Comment (shira-g) [reply nlp_architect/data/sequential_tagging.py]: moved to data/utils.py

File: nlp_architect/models/absa/train/acquire_terms.py [hunk 8]
```diff
@@ -21,7 +21,7 @@
 from tqdm import tqdm
 
 from nlp_architect.models.absa import TRAIN_LEXICONS, LEXICONS_OUT
-from nlp_architect.models.absa import TRAIN_OUT, GENERIC_OP_LEX
+from nlp_architect.models.absa import GENERIC_OP_LEX
```
[the remark this answers is not in the document]
Comment (shira-g) [reply nlp_architect/models/absa/train/acquire_terms.py]: it's a pyling fix

File: nlp_architect/data/utils.py [hunk 9]
```diff
@@ -154,3 +148,39 @@ def sample_label_unlabeled(samples: List[InputExample], no_labeled: int, no_unla
     label_samples = [samples[i] for i in labeled_indices]
     unlabel_samples = [samples[i] for i in unlabeled_indices]
     return label_samples, unlabel_samples
+
+
+def split_column_dataset(
+        m: int, n: int, out_folder, dataset, m_filename='m.txt', n_filename='n.txt', tag_col=-1):
```
Comment (peteriz) [inline nlp_architect/data/utils.py]: * do not specify output file names, let the user choose
* please rename m/n to a more semantic name

File: nlp_architect/data/utils.py [hunk 10]
```diff
@@ -154,3 +148,39 @@ def sample_label_unlabeled(samples: List[InputExample], no_labeled: int, no_unla
...
+    num_of_examples = len(lines)
+    assert m + n <= num_of_examples and m > 0 and n > 0
+    selected_lines = random.sample(lines, m + n)
+    m_data = selected_lines[:m]
+    n_data = selected_lines[m:]
+    write_column_tagged_file(out_folder + os.sep + m_filename, m_data)
+    write_column_tagged_file(out_folder + os.sep + n_filename, n_data)
+
+
+def count_examples(file):
```
Comment (peteriz) [inline nlp_architect/data/utils.py]: is this method used in the library? if only in tests please move it there

File: tests/cdc/test_utils.py [hunk 11]
```diff
@@ -226,3 +227,16 @@ def get_wordnet_mentions():
         mentions.append(MentionData.read_json_mention_data_line(json))
 
     return mentions
+
+
+def count_examples(file):
```
Comment (peteriz) [inline tests/cdc/test_utils.py]: why is this here? `cdc` is Alon's package.

File: tests/test_data_utils.py [hunk 12]
```diff
@@ -16,7 +16,8 @@
 
 import math
 import os
-from nlp_architect.data.utils import split_column_dataset, count_examples
+from nlp_architect.data.utils import split_column_dataset
+from tests.cdc.test_utils import count_examples
```
Comment (peteriz) [inline tests/test_data_utils.py]: same comment as above
